### PR TITLE
[Diffusion] Optimize Qwen-Image-Edit attention on Hopper

### PR DESCRIPTION
**Pull request by BBuf**
## Motivation

Qwen-Image-Edit's Hopper profile contains repeated QK normalization/RoPE and joint-QKV concatenations in the DiT, plus 96 separate SDPA window calls per window-attention layer in the vision encoder. Reuse the existing Qwen QKV epilogue on Hopper and process packed vision windows through PyTorch's public varlen attention API.

On one H200 with main's PyTorch 2.13 dependencies, lossless eager engine E2E falls from **32.8171 s to 32.3141 s (1.533%)**. All saved before/after images are byte-identical within each quality mode. This is a modest combined full-model gain; the much larger microbenchmark ratios do not represent E2E speedups.

## Modifications

- Enable the existing BF16, batch-one, head-dimension-128 QKV epilogue for eligible unquantized SM90 Qwen attention. Support independent image/text token strides; retain mask, shard/padding, layout and alignment fallbacks and the existing environment opt-out.
- Vectorize adjacent RoPE-cache loads and paired V-head copies, and move Q/K/V dispatch into the grid's second dimension. Explicit `fmaf`/`fmul` preserves the original RoPE contraction and intermediate rounding. Odd head counts retain a smaller final copy.
- Use public `torch.nn.attention.varlen.varlen_attn` for eligible Hopper Flash SDPA inference with packed windows. No token reordering or padding. CPU/HIP/MPS, other SDPA subclasses, compilation, gradients, dropout, missing API, and unsupported inputs retain the existing path. The new SDPA dispatch is SM90-only; the shared QKV kernel also serves the existing SM100 path, whose B200 regression now passes CI (14 QKV epilogue tests).
- Update the kernel inventory's SM90 capability and extend the existing kernel, model-path and registry tests.

## Accuracy Tests

Rehashed **16 formal PNGs**: four A-B-B-A runs × two quality modes × two Torch versions. Within each mode, A/A, B/B and A/B are byte-identical, with SSIM 1 and maximum pixel error 0. Actual output resolution is 1024×1024.

The contact sheet shows baseline and candidate in columns, lossless and high in rows:

Actual Qwen-Image-Edit before and after outputs

| Quality | Actual baseline PNG | Actual candidate PNG |
| --- | --- | --- |
| Lossless | [Before](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/lossless-baseline.png) | [After](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/lossless-candidate.png) |
| High | [Before](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/high-baseline.png) | [After](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/high-candidate.png) |

**Use lossless for this recipe.** High already changes this input visibly: high versus lossless has SSIM **0.830311**, PSNR 17.3808 dB and maximum difference 255/255, identically for baseline and candidate. The high timing below documents compatibility with that existing mode; it is not a lossless-equivalent result or a new quality optimization.

[Torch 2.13 image comparisons](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch213-image-quality.json), [Torch 2.11 comparisons](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch211-image-quality.json), [all formal hashes, source checks and selected-GPU samples](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/artifact-gpu-audit.json).

- **35 H200 tests passed on each of Torch 2.11.0+cu130 and 2.13.0+cu130**. Coverage includes QKV exactness, separate and packed strides, odd head counts, misaligned cache/token-stride fallbacks, masked attention, packed SDPA BF16/FP16, causal/ragged/empty/single windows, strided V, and training/dropout/missing-API fallbacks.
- **28 CPU registry tests passed** at the final PR head. All nine changed files passed pre-commit and `git diff --check`.
- Production microbenchmarks check six QKV cases and fourteen SDPA cases exactly. The existing B200 CI lane passes all 14 QKV epilogue tests, including the production 8,152/1,365-token shape, odd heads, and mask/shard/alignment guards. A Hopper lane is added.

[Final checks](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/final-checks.json), [2.11 test log](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch211-tests.log), [2.13 test log](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch213-tests.log), [registry test log](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/registry-tests.log).

## Speed Tests and Profiling

Native `QwenImageEditPipeline`, `Qwen/Qwen-Image-Edit@ac7f9318f633fc4b5778c59367c8128225f1e3de`; one H200, BF16, 1024×1024, 50 steps, guidance 4, embedded guidance 6, seed 42, official Qwen Edit input, prompt `Make the cat wear a red hat.` Automatic residency in both arms, including the existing text/image encoder layerwise offload. Eager only, no torch.compile. Native BCG probes were disabled for both qualities, so no BCG fallback timings are counted.

A is main `482e9f257bb9d9ac57c2245c93768a96ec70edbe`. Lossless measurements and microbenchmarks use candidate `60def09d35c8912db76144e7ca6691dca554bc94`; final head `7df1e941ca7782d9a964a98a9720d4ddd1cd7fa4` adds only registry capability metadata, README and CPU tests. The four runtime files are byte-identical ([audit](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/runtime-equivalence.json)). Primary high and paired profiles measured the actual final head. Runtime import paths and perf-dump commits were verified.

Each row uses fresh processes in A-B-B-A order on the same selected GPU. No rows were discarded. Primary environment: PyTorch 2.13.0+cu130, sglang-kernel 0.4.6.post1, driver 595.71.05. It was installed in a private venv; the separate legacy 2.11 environment was unchanged.

| Primary Torch 2.13 eager | A1 / A2 engine E2E | B1 / B2 engine E2E | Mean A → B | Reduction |
| --- | ---: | ---: | ---: | ---: |
| Lossless | 32.805944 / 32.828280 s | 32.300505 / 32.327791 s | 32.817112 → 32.314148 s | **1.533%** |
| High, existing quality tradeoff | 32.290089 / 32.280895 s | 31.610836 / 31.596315 s | 32.285492 → 31.603576 s | **2.112%** |

| Environment / quality | Mean denoising A → B | Denoising reduction | Mean engine E2E A → B | E2E reduction |
| --- | ---: | ---: | ---: | ---: |
| Torch 2.13 lossless | 31.821445 → 31.486249 s | 1.053% | 32.817112 → 32.314148 s | 1.533% |
| Torch 2.13 high | 31.296227 → 30.783327 s | 1.639% | 32.285492 → 31.603576 s | 2.112% |
| Torch 2.11 lossless | 31.936158 → 31.592483 s | 1.076% | 32.930535 → 32.441389 s | 1.485% |
| Torch 2.11 high | 31.378467 → 30.879241 s | 1.591% | 32.385151 → 31.710122 s | 2.084% |

The legacy lossless result is below the campaign's 1.5% acceptance threshold; it is not rounded up or pooled with 2.13. Only one input/seed and two measurements per arm are reported, so this does not establish a universal 1.5% improvement across Qwen models or inputs.

Engine E2E is synchronized stage `total_duration_ms`, excluding model loading, request warmup and PNG saving. Profile-export wall times are excluded. In unprofiled primary lossless runs, image encoding is 792.848 → 632.484 ms and denoising saves 335.196 ms. Selected GPUs were idle at every cell entry. Ten-second samples, including two legacy NVML query timeouts, are preserved; they do not prove continuous host exclusivity.

[Primary lossless ABBA](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch213-lossless-abba.json), [primary high ABBA](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch213-high-abba.json), [legacy lossless ABBA](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch211-lossless-abba.json), [legacy high ABBA](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/torch211-high-abba.json).

### Kernel and profile evidence

The initial trace identified actual DiT shapes of 8,152 image tokens and 1,365/1,358 text tokens, 24 heads, head dimension 128. Vision attention has 5,400 tokens over 96 ragged windows, 16 heads, dimension 80. Measurements below are cumulative GPU kernel durations from separate tiny Torch 2.11 traces, excluding CPU launch gaps; they are not full request timings.

| Production operation | Baseline kernels / device time | Candidate kernels / device time |
| --- | ---: | ---: |
| QK norm + RoPE + joint QKV, 8,152 + 1,365 tokens | 7 / 179.617 µs | 1 / 109.023 µs |
| Packed vision SDPA, 5,400 tokens / 96 windows | 96 / 517.984 µs | 2 / 38.240 µs |

The two SDPA kernels are FlashAttention and an auxiliary zero-RNG fill; dropout is zero and no random numbers are consumed. Copy events are separate from kernel counts. QKV input restoration is outside the timed interval; the reference's packed-input materialization cost is conservatively excluded.

NCU localized 7,309,056 excess L2 sectors to four stride-two scalar RoPE-cache loads. Vector loads remove those sectors, but alone do not improve duration (121.248 → 121.760 µs). Paired V-head copies alone regress to 131.840 µs; hoisting dispatch/index work into `grid.y` reduces it to **112.544 µs**, with 63,691,368 executed instructions and 93.204% occupancy. Explicit FMA operations are necessary to retain bit-exact arithmetic. These are NCU diagnostics with clock/cache control disabled, not E2E measurements.

The final matched Torch 2.13 full-stage captures contain two denoising steps each. Overlapping image-encoding GPU annotations are unioned before counting each kernel once; no boundary-crossing kernels were found.

| Rank-0 stage, two-step request | A kernels → B kernels | A cumulative GPU time → B |
| --- | ---: | ---: |
| Denoising | 7,113 → 5,673 | 1,233.223 → 1,221.776 ms |
| Image encoding | 11,042 → 5,778 | 190.449 → 164.337 ms |

Other stage kernel counts are unchanged. The vision reduction is 56 window-attention invocations × (96 − 2) = 5,264 launches. These short instrumented requests establish the changed paths; the full 50-step unprofiled ABBA establishes E2E.

[QKV microbenchmark](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/qkv-microbenchmark.json), [SDPA microbenchmark](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/sdpa-microbenchmark.json), [NCU iterations](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/ncu-comparison.json), [paired profile and trace hashes](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/profile-comparison.json), [profile aggregation script](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/summarize_qwen_attention_profile.py).

### Reproduce

Run the same command on the stated baseline and candidate, in A-B-B-A order, with the same idle H200 and environment:

```bash
CUDA_VISIBLE_DEVICES=0 SGLANG_DIFFUSION_SYNC_STAGE_PROFILING=1 sglang generate \
  --backend sglang --model-path Qwen/Qwen-Image-Edit \
  --revision ac7f9318f633fc4b5778c59367c8128225f1e3de \
  --num-gpus 1 --performance-mode auto --enable-torch-compile false \
  --quality lossless --warmup-mode request --width 1024 --height 1024 \
  --num-inference-steps 50 --guidance-scale 4 \
  --seed 42 --prompt 'Make the cat wear a red hat.' \
  --image-path https://github.com/lm-sys/lm-sys.github.io/releases/download/test/TI2I_Qwen_Image_Edit_Input.jpg \
  --save-output --output-path outputs/qwen-edit --perf-dump-path qwen-edit.perf.json
```

For high, change only `--quality high` in both arms and retain its quality caveat. Capture full stages in separate diagnostic requests with two inference steps, `--profile` and `--profile-all-stages`; exclude their wall times. Exact executed [QKV harness](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/bench_qwen_qkv_hopper_production.py) and [SDPA harness](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/bench_native_varlen_final.py) are included; their output directory and commit assertions record the original experiment.

Task-owned checkpoint caches on both H200 hosts were deleted after model validation: 57,722,203,391 and 57,722,164,005 logical bytes before cleanup, zero files/weights/bytes afterward. These are logical cache counts, not unique physical storage freed. [First ledger](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/ion9-cache-cleanup.jsonl), [second ledger](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/ion8-cache-cleanup.jsonl), [independent zero-residual check](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/cleanup-verification.json), [media manifest](https://raw.githubusercontent.com/BBuf/sglang/pr-media/diffusion-prs/qwen-edit-attention-h200/manifest.json).

## Checklist

- [x] Format code and pass pre-commit.
- [x] Extend registered tests and run focused H200/CPU checks.
- [x] Update kernel inventory documentation.
- [x] Provide accuracy, kernel, profile, E2E and actual output comparisons.
- [x] Follow SGLang code style.
- [x] Existing [B200 kernel regression CI](https://github.com/sgl-project/sglang/actions/runs/34296369242/job/102293942926): all 14 QKV epilogue tests pass on `b200-cirrascale1-0123`. [B200 diffusion CI](https://github.com/sgl-project/sglang/actions/runs/34296369242/job/102293942990) also passes.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34661251240](https://github.com/sgl-project/sglang/actions/runs/34661251240)
Latest PR Test (Extra): :white_check_mark: [Run #34712119649](https://github.com/sgl-project/sglang/actions/runs/34712119649)
Latest PR Test (AMD ROCm 10): :x: [Run #34661251381](https://github.com/sgl-project/sglang/actions/runs/34661251381)

### CODE REVIEW

**Review by mickqian — COMMENTED** [review]
CI follow-up on 3f3a0d919436e779955774b7797d55910c9c1c7e after the bounded rerun (Base attempt 4):

| Check | Evidence |
|---|---|
| [2-GPU shard 0](https://github.com/sgl-project/sglang/actions/runs/34459909373/job/103408147773) | Wan2.2 TeaCache again stops at warmup 0/3 until startup timeout; the shard later reaches the 45-minute job limit. Wan2.2 I2V and LoRA generation do pass. |
| [2-GPU shard 1](https://github.com/sgl-project/sglang/actions/runs/34459909373/job/103408147865) | MiniMax H3 passes; Wan2.2 T2V and Wan2.1 I2V 480p remain at warmup 0/3, and the shard reaches its 45-minute limit. |
| [2-GPU shard 2](https://github.com/sgl-project/sglang/actions/runs/34459909373/job/103408147739) | Wan2.1 I2V LoRA reports `Server not ready within 1200.0s`, then its retry is interrupted by the job limit. There is also a Qwen-image average-denoise failure: 233.4772 ms > 231.4800 ms limit. LTX-2.3 TI2V initially reports 325.9838 ms > 300.1920 ms but passes its internal retry. |
| [Base-B 2-GPU-large (4)](https://github.com/sgl-project/sglang/actions/runs/34459909373/job/103408146413) | Health-check fast-fail caused by diffusion-coverage-check; not an independent LLM failure. |

The startup stalls persisted across two completed rerun attempts on different runners, so another blind whole-shard rerun is not a diagnosis. This head does not include the IPC JIT recovery fix #39034 (358c163250ad3b1f62939b01ce1314a0a31a0365), but warmup logs alone cannot establish that every stall has that cause. A blocked-worker stack (or a controlled same-configuration comparison of the affected IPC initialization path) is needed before attributing these failures. The Qwen performance assertion must be evaluated separately; do not relax its threshold or assume the IPC fix addresses it.

No further rerun or branch synchronization was issued in this follow-up.

**Comment by BBuf** [comment]
https://github.com/sgl-project/sglang/actions/runs/34661251240/job/103484060640?pr=38584
